src/Xamarin.Android.Build.Tasks/Utilities/AndroidResource.cs

```diff
@@ -3,101 +3,121 @@
 using System.IO;
 using System.Linq;
 using System.Text.RegularExpressions;
+using System.Xml;
 using System.Xml.Linq;
 using System.Xml.XPath;
+using System.Text;
 
 namespace Monodroid {
-	static class AndroidResource {
-		
+	public static class AndroidResource {
+
+		public static void UpdateXmlResource (string res, Stream source, Stream dest, Dictionary<string, string> acwMap, IEnumerable<string> additionalDirectories = null, bool leaveOpen = false)
+		{
+			string path = string.Empty;
+			using (var writer = new LinePreservedXmlWriter (new StreamWriter (dest, System.Text.Encoding.UTF8, 4000, leaveOpen: leaveOpen))) {
+				var reader = XmlReader.Create (source);
+				while (reader.Read ()) {
+					if (reader.NodeType == XmlNodeType.Element) {
+						path += "/" + reader.LocalName;
+					}
+					WriteShallowNode (reader, writer, path, acwMap, res, additionalDirectories);
+					if (reader.NodeType == XmlNodeType.Element) {
+						if (reader.HasAttributes) {
+							var inTransition = (reader.Name == "transition");
+							var inFragment = (reader.Name == "fragment");
+							var parentLocalName = reader.LocalName;
+							var parentNamespace = reader.NamespaceURI;
+							for (int attInd = 0; attInd < reader.AttributeCount; attInd++) {
+								reader.MoveToAttribute (attInd);
+
+								string value = reader.Value;
+								Match m = r.Match (value);
+								if (m.Success)
+									value = TryLowercaseValue (value, res, additionalDirectories);
+								if (reader.Prefix == "xmlns") {
+									writer.WriteAttributeString (reader.Name, value);
+									continue;
+								}
+								TryFixResAuto (reader, acwMap, ref value);
+								if (inFragment)
+									TryFixFragment (reader, acwMap, ref value);
+								TryFixCustomClassAttribute (reader, inTransition, acwMap, ref value);
+
+								if (reader.NamespaceURI != android && !(reader.LocalName == "layout" && reader.NamespaceURI == string.Empty &&
+										parentLocalName == "include" && parentNamespace == string.Empty)) {
+									writer.WriteAttributeString (reader.Name, value);
+									continue;
+								}
+								writer.WriteAttributeString (reader.Name, value);
+							}
+							reader.MoveToElement ();
+						}
+						if (reader.IsEmptyElement) {
+							writer.WriteEndElement ();
+							path = path.Substring (0, path.LastIndexOf ('/'));
+						}
+					}
+					if (reader.NodeType == XmlNodeType.EndElement) {
+						path = path.Substring (0, path.LastIndexOf ('/'));
+					}
+				}
```

Comment:
`path` above could be a `StringBuilder` - you would save on allocation (each time you do `+= '/'` a new string is created) and you always know the position of the last `/` - so you'd save on calls to `LastIndexOf`. Code would get slightly more complicated but, IMO, faster and leaner